cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto

```diff
@@ -54,11 +54,17 @@ service PackagesService {
   //     put: "/core/packages/v1alpha1/installedpackages"
   //   };
   // }
-  // rpc DeleteInstalledPackage(DeleteInstalledPackageRequest) returns (DeleteInstalledPackageResponse) {
-  //   option (google.api.http) = {
-  //     delete: "/core/packages/v1alpha1/installedpackages"
-  //   };
-  // }
+  rpc DeleteInstalledPackage(DeleteInstalledPackageRequest) returns (DeleteInstalledPackageResponse) {
+    option (google.api.http) = {
+      // Example:
+      // /core/packages/v1alpha1/installedpackages/plugin/helm.packages/v1alpha1/ns/my-ns/release-identifier
+      delete: "/core/packages/v1alpha1/installedpackages/plugin/{installed_package_ref.plugin.name}/{installed_package_ref.plugin.version}/ns/{installed_package_ref.context.namespace}/{installed_package_ref.identifier}"
```

Comment:
What I wanted to do here was just `/core/packages/v1alpha1/installedpackages/{identifier built by b64 of the installed package ref}` but it's probably not a good idea (wouldn't be unique) and it's not possible with the grpc transcoding.

So instead of `/installedpackages/{identifier of some sort}` we have a textual identifier built from the plugin/namespace/installed package identifier.

For the plugins (as opposed to the core), the identifier is just the namespace and identifier (since we know the plugin).

Reply:
I'd rather go with the "long" identifier solution, albeit more prone to enumeration attacks, it follows the REST spirit. 

I also was going to mention we should add the `cluster` to the path. Reading your comment below... mmm, you're right in that we may not want to make the cluster explicit... but I don't know if it's a good idea to use the message body in a DELETE msg.

Do you know if we can have two HTTP paths linked to the same grpc operation? If so, we could have one with cluster and another one without.

If not... adding a body to a DELETE method is really against the protocol. Whereas is not disallowed... the msg body of a DELETE should be ignored.
That's why there are some API clients (Java Jersey I think?, some Android httpclients and several SaaS systems) that just don't support bodies in DELETEs. 
So... even if technically possible... it could translate into UX issues for developers (which wouldn't be a big deal if it weren't for it being aimed at being consumed by 3rd party people)

Let's see what Greg thinks, but my -1 to use bodies for now :S

Reply:
> Do you know if we can have two HTTP paths linked to the same grpc operation? If so, we could have one with cluster and another one without.

Let me try - that'd be a much better solution if we can.

> 
> If not... adding a body to a DELETE method is really against the protocol. Whereas is not disallowed... the msg body of a DELETE should be ignored.
> That's why there are some API clients (Java Jersey I think?, some Android httpclients and several SaaS systems) that just don't support bodies in DELETEs.
> So... even if technically possible... it could translate into UX issues for developers (which wouldn't be a big deal if it weren't for it being aimed at being consumed by 3rd party people)

Yeah - I'm totally happy to update to remove it. My only concern was having the cluster in the URL endpoints when certain scenarios of how our API could be used in the future might actively not want multi-cluster support. But it sounds like it'd be worse to have the body in delete than to have a redundant `/c/default/` in the URL. I'll check as above first if we can define multiple paths (I know we can define multiple paths, just unsure whether that can be for the same method, doubt it), but if not, will update removing the body option.

Thanks Antonio.

Reply:
Tried with:

```diff
--- a/cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto
+++ b/cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto
@@ -60,6 +60,7 @@ service PackagesService {
       // Example:
       // /core/packages/v1alpha1/installedpackages/plugin/helm.packages/v1alpha1/c/default/ns/my-ns/release-identifier
       delete: "/core/packages/v1alpha1/installedpackages/plugin/{installed_package_ref.plugin.name}/{installed_package_ref.plugin.version}/c/{installed_package_ref.context.cluster}/ns/{installed_package_ref.context.namespace}/{installed_package_ref.identifier}"
+      delete: "/core/packages/v1alpha1/installedpackages/plugin/{installed_package_ref.plugin.name}/{installed_package_ref.plugin.version}/ns/{installed_package_ref.context.namespace}/{installed_package_ref.identifier}"
     };
   }
 }
```

but

```bash
buf generate

proto/kubeappsapis/core/packages/v1alpha1/packages.proto:63:7:method kubeappsapis.core.packages.v1alpha1.PackagesService.DeleteInstalledPackage: option (google.api.http): non-repeated option field delete already set
```

So for now I've just pushed with the url including the cluster (and no more body on delete :) ).

Reply:
I don't quite understand. DELETE is not so unique. We have several GET operations which cannot have bodies either. So buf happily generates request query parameters for everything. Why would DELETE be so different? See attachment
<img width="1440" alt="Screen Shot 2021-09-16 at 6 35 02 PM" src="https://user-images.githubusercontent.com/74032303/133709753-7a89e4b3-fa32-4a1b-b527-e6538b57e797.png">
is it the case you're trying to accomplish 2 things in one PR:
1) add stubs for DELETE and
2) change the way DELETE is transcoded for grpc?

If so, I would break this into 2 separate PRs:
1) add stubs for DELETE and
2) change the way ALL operations that need so are transcoded so we have a consistent pattern. Not just one sore thumb sticking out

Reply:
GET params can be transcoded from URI params without issue to populate the mesages. DELETE is different enough that the tooling does not allow DELETE with a body by default. See Antonio's comment for more info.



> Is it the case you're trying to accomplish 2 things in one PR:
> 
>     1. add stubs for DELETE and
> 
>     2. change the way DELETE is transcoded for grpc?

I didn't change it, we weren't doing it at all. I was planning to just add the stubs for DELETE but both grpc-gateway *and* openapi refused to allow a body without overriding explicitly (as I'd done originally, you can see it in a previous commit). So I wanted to discuss what would be best for DELETE specifically here and be able to follow up with a PR to propose adding more REST-like transcoding for the other msgs

> 
> 
> If so, may I suggest you split this into 2 separate PRs:
> 
>     1. add stubs for DELETE consistent with the rest of the APIs. So DELETE isn't one sore thumb sticking out and once you land that then

Happy to do so if that makes things easier. I thought it was worthwhile looking at DELETE in particular due to the above warnings that the tooling adds about using a body with DELETE... it seems it is special, but I'll let Antonio (he has a PhD's worth of REST/OpenAPI knowledge) chime in.
 
> 
>     2. change the way ALL operations that need so are transcoded so we have a consistent pattern. We can have a productive discussion then that covers all use cases

Yes, that's the plan (as mentioned in the description), to follow up with a PR proposing the rest if we agree here on the structure for DELETE (cluster/namespace/etc.).

Reply:
> GET params can be transcoded from URI params without issue to populate the mesages. DELETE is different enough that the tooling does not allow DELETE with a body by default. See Antonio's comment for more info.
> 
> > Is it the case you're trying to accomplish 2 things in one PR:
> > ```
> > 1. add stubs for DELETE and
> > 
> > 2. change the way DELETE is transcoded for grpc?
> > ```
> 
> I didn't change it, we weren't doing it at all. I was planning to just add the stubs for DELETE but both grpc-gateway _and_ openapi refused to allow a body without overriding explicitly (as I'd done originally, you can see it in a previous commit). So I wanted to discuss what would be best for DELETE specifically here and be able to follow up with a PR to propose adding more REST-like transcoding for the other msgs
> 
> > If so, may I suggest you split this into 2 separate PRs:
> > ```
> > 1. add stubs for DELETE consistent with the rest of the APIs. So DELETE isn't one sore thumb sticking out and once you land that then
> > ```
> 
> Happy to do so if that makes things easier. I thought it was worthwhile looking at DELETE in particular due to the above warnings that the tooling adds about using a body with DELETE... it seems it is special, but I'll let Antonio (he has a PhD's worth of REST/OpenAPI knowledge) chime in.
> 
> > ```
> > 2. change the way ALL operations that need so are transcoded so we have a consistent pattern. We can have a productive discussion then that covers all use cases
> > ```
> 
> Yes, that's the plan (as mentioned in the description), to follow up with a PR proposing the rest if we agree here on the structure for DELETE (cluster/namespace/etc.).

I read Antonio's comment. Still confused why DELETE is different than a GET in that respect. DELETE with a body is a non-starter. Same as GET. I have known this and operated with this knowledge for possibly at least 8-10 years, ever since my introduction to REST. Having any kind of tooling that allows a body with DELETE should have been a non-starter. I propose we treat DELETE same as GET in this PR, that is have the tooling generate query params for inputs. Then in a different PR, lets discuss how to make the world better :-)

Reply:
Sorry, I'd misread your comment and thought you wanted it consistent with post etc., but see it was GET. Great, so everyone is in agreement about not using body, and Greg is keen that we don't propose a better URL for delete in this PR, but instead use the same and update all in one PR, which makes sense. I've pushed that.